server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java

```diff
@@ -148,15 +159,23 @@ private static void notImplemented(String input) {
                              messageMapper.findInMailbox(mailbox, MessageRange.one(messageId.getUid()), MessageMapper.FetchType.Full, 1),
                              mailboxPath
                              );
-         })
-                .andThen(this::iteratorToStream);
+                })
+                .andThen(pair -> iteratorToStream(pair, mailboxSession));
     }
     
-    private Stream<Pair<MailboxMessage, MailboxPath>> iteratorToStream(Pair<Iterator<MailboxMessage>, MailboxPath> value) {
+    private Stream<Triplet<MailboxMessage, List<Attachment>, MailboxPath>> iteratorToStream(Pair<Iterator<MailboxMessage>, MailboxPath> value, MailboxSession mailboxSession) {
         Iterable<MailboxMessage> iterable = () -> value.getValue0();
         Stream<MailboxMessage> targetStream = StreamSupport.stream(iterable.spliterator(), false);
         
         MailboxPath mailboxPath = value.getValue1();
-        return targetStream.map(x -> Pair.with(x, mailboxPath));
+        return targetStream.map(Throwing
+                .function((MailboxMessage message) -> Triplet.with(message, retrieveAttachments(message.getAttachmentsIds(), mailboxSession), mailboxPath)));
+    }
+
+    private List<Attachment> retrieveAttachments(List<AttachmentId> attachmentsIds, MailboxSession mailboxSession) throws MailboxException {
+        AttachmentMapper attachmentMapper = attachmentMapperFactory.getAttachmentMapper(mailboxSession);
```

Comment:
you take a mapper for every single message, you could curry the function (see https://en.wikipedia.org/wiki/Currying )